LoRaWAN: Remedy for issue #7230

hasnainvirk opened the pull request:
### Description

Link for the original issue https://github.com/ARMmbed/mbed-os/issues/7230
Unfortunately, a refactor in the past slightly changed the behaviour of re-connection and we stopped using LORAWAN_CONNECT_IN_PROGRESS return code. This was added in the first place to help the application know that the stack is waiting for a JoinAccept from the network server. There was a confusion created because of this mistake regarding re-connection. We have now devised clear rules about what can be done and how its done. Here are the rules which are now added to the API documentation:

For ABP:  If everything goes well, LORAWAN_STATUS_OK is returned for first call followed by a 
                'CONNECTED' event. Otherwise a negative error code. Any subsequent call will return 
                LORAWAN_STATUS_ALREADY_CONNECTED and no event follows.

For OTAA: When a JoinRequest is sent, LORAWAN_STATUS_CONNECT_IN_PROGRESS is returned followed 
                 by a 'CONNECTED' event when the JoinAccept is received. Otherwise a negative error code is 
                 returned. Any subsequent call will return LORAWAN_STATUS_ALREADY_CONNECTED and no 
                 event follows.

### TARGET release
Mbed OS 5.10





### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [X] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

Commit:
LoRaWAN: Handling re-joining when already Joined

This is a remedy for the issue #7230.
While the device is joining, LORAWAN_STATUS_CONNECT_IN_PROGRESS is returned.
However, if the device is already joined, we will return LORAWAN_STATUS_ALREADY_CONNECTED.

Commit:
Updating docs

API documentation is updated to clear how the connection related return codes will
work from now on.

hasnainvirk commented:
@kivaisan @kjbracey-arm Please review.

kjbracey commented:
> For OTAA: When a JoinRequest is sent, LORAWAN_STATUS_CONNECT_IN_PROGRESS is returned followed
> by a 'CONNECTED' event when the JoinAccept is received. Otherwise a negative error code is
> returned. 

Otherwise? Does that mean if it doesn't send a JoinRequest? I guess so, but the phrasing is a bit convoluted here - "When this..., followed by X when... Otherwise..."

I may have raised this before - POSIX has an extra error code, which distinguish the two cases of "this connect request is now in progress" (EINPROGRESS) and "there was a connect already in progress, so I'm ignoring this call" (EALREADY). (And both are distinct from "this is connected, so I'm ignoring you" (EISCONN)".

Is it worth distinguishing that - what code do you currently get if this is called between JoinRequest and JoinAccept event?

Commit:
Streamlining connect() API with posix like retcodes

For ABP: First call to connect() or connect(params) will return LORAWAN_STATUS_OK
         and a CONNECTED event will be sent. Any subsequent call will return
         LORAWAN_STATUS_ALREADY_CONNECTED (posix EISCONN) and no event is generated.

FOR OTAA: First call to connect() or connect(params) will return LORAWAN_STATUS_CONNECT_IN_PROGRESS
          and a CONNECTED event will be sent whenever the JoinAccept is received. If the application
          calls connect again before receiving the CONNECTED event, LORAWAN_STATUS_BUSY will be returned.
          After the CONNECTED event is dispatched, any subsequent call to connect() or connect(params) API
          will be returned with LORWAN_STATUS_ALREADY_CONNECTED.

No new parameters are accepted after the first call. The application must disconnect before making
a connect() call with new parameters.

hasnainvirk commented:
@kjbracey-arm Please review again.

hasnainvirk commented:
@kjbracey-arm @cmonr @0xc0170 Please review.

hasnainvirk commented:
@0xc0170 Needs CI here.

cmonr commented:
/morph build

mattbrown015 commented:
@hasnainvirk When there is something wrong with the LoRaWAN session parameters the 'JOIN_FAILURE' event occurs after the first join request transmission.  This surprised me, I expected the 'JOIN_FAILURE' after 12 join attempts.

Immediately after the 'JOIN_FAILURE' I see another transmission but then that's it.  In other words, I can't see the 12 join attempts that I would expect.

I thought I'd seen a discussion about the expected behaviour of 'JOIN_FAILURE' somewhere but I can't find it now.

I'm using [38744b9](https://github.com/ARMmbed/mbed-os/commit/38744b9e6877f9e7b0cabffd17423a1a22c4a6ea).

mattbrown015 commented:
I think the problem is caused by [LoRaWAN: Handling re-joining when already Joined](https://github.com/ARMmbed/mbed-os/commit/51f92b0efdc60f8eec59cfb3cc30bd0aa188c1e4) where the return value of `LoRaMac::send_join_request()` changes from `LORAWAN_STATUS_OK` to `LORAWAN_STATUS_CONNECT_IN_PROGRESS`.

The following change meant the join got retried again:

    diff --git a/features/lorawan/lorastack/mac/LoRaMac.cpp b/features/lorawan/lorastack/mac/LoRaMac.cpp
    index 2a925744d..6d7ed946b 100644
    --- a/features/lorawan/lorastack/mac/LoRaMac.cpp
    +++ b/features/lorawan/lorastack/mac/LoRaMac.cpp
    @@ -772,7 +772,7 @@ bool LoRaMac::continue_joining_process()
         }

         // Schedule a retry
    -    if (handle_retransmission() != LORAWAN_STATUS_OK) {
    +    if (handle_retransmission() != LORAWAN_STATUS_CONNECT_IN_PROGRESS) {
             return false;
         }

hasnainvirk commented:
@mattbrown015 You are right.  My local branch had the code correctly as you have shown. For some reason my PR still had a check for OK status.

mattbrown015 commented:
@hasnainvirk Great! Relatively easy to sort out then. :-)

hasnainvirk commented:
@mattbrown015 I have opened a PR for the fix. handle_retransmission() is shared between both Join and Data paths that's why the confusion happened. https://github.com/ARMmbed/mbed-os/pull/7520

0xc0170 commented:
@mattbrown015 thanks for bringing this up
